### PR TITLE
Allow partially moved values in match

### LINKED ISSUE #99180: MIR building optimizes away some place expressions around "_" patterns even with -Zmir-opt-level=0
Consider the following code -- the comments say it all:
```rust
fn main() {
    let ptr = 32 as *const u8;
    
    // This is UB, and Miri finds it:
    //unsafe { let _ptr = std::ptr::addr_of!(*ptr); }
    
    // So this should also be UB, but Miri says nothing:
    unsafe { let _ = *ptr; }
    
    // And same for this:
    unsafe {
        match *ptr { _ => {} }
    }
}
```
Here is the mir-opt-level=0 MIR for this:
```
fn main() -> () {
    let mut _0: ();                      // return place in scope 0 at place.rs:1:11: 1:11
    let _1: *const u8;                   // in scope 0 at place.rs:2:9: 2:12
    let _2: ();                          // in scope 0 at place.rs:8:5: 8:29
    scope 1 {
        debug ptr => _1;                 // in scope 1 at place.rs:2:9: 2:12
        scope 2 {
            scope 3 {
            }
        }
        scope 4 {
        }
    }

    bb0: {
        StorageLive(_1);                 // scope 0 at place.rs:2:9: 2:12
        _1 = const 32_usize as *const u8 (PointerFromExposedAddress); // scope 0 at place.rs:2:15: 2:30
        StorageLive(_2);                 // scope 1 at place.rs:8:5: 8:29
        _2 = const ();                   // scope 2 at place.rs:8:5: 8:29
        StorageDead(_2);                 // scope 1 at place.rs:8:28: 8:29
        _0 = const ();                   // scope 4 at place.rs:12:27: 12:29
        StorageDead(_1);                 // scope 0 at place.rs:14:1: 14:2
        return;                          // scope 0 at place.rs:14:2: 14:2
    }
}
```
Currently, MIR building "loses" the entire place expression somewhere, *even though that place expression could cause UB*. IMO that is a bug. Can we change MIR building to preserve the place expressions?

We could have a `EvalPlace` MIR statement that just computes a place expression and then does nothing with the result. (To be clear, there should be *no* place-to-value conversion, i.e., no actual load from the `*ptr` place. But the place expression itself should be evaluated.) We *do* have `FakeRead`, which works on a place, but (a) its docs talk about "inspecting constants and discriminant values" so that is doing more than just evaluating the place expression, and (b) it is "disallowed after drop elaboration".

Cc @rust-lang/wg-mir-opt in lieu of a "MIR building wg".
Also I think this is related to THIR unsafety checking; MIR unsafety checking does not even complain about some of these `*ptr` because they get erased before reaching MIR, but in THIR they would still be visible.
Also related: https://github.com/rust-lang/rust/issues/79735 https://github.com/rust-lang/unsafe-code-guidelines/issues/261

### PR DESCRIPTION
**Pull request by cjgillot**
This PR attempts to unify the behaviour between `let _ = PLACE`, `let _: TY = PLACE;` and `match PLACE { _ => {} }`.
The logical conclusion is that the `match` version should not check for uninitialised places nor check that borrows are still live.

The `match PLACE {}` case is handled by keeping a `FakeRead` in the unreachable fallback case to verify that `PLACE` has a legal value.

Schematically, `match PLACE { arms }` in surface rust becomes in MIR:
```rust
PlaceMention(PLACE)
match PLACE {
  // Decision tree for the explicit arms
  arms,
  // An extra fallback arm
  _ => {
    FakeRead(ForMatchedPlace, PLACE);
    unreachable
  }
}
```

`match *borrow { _ => {} }` continues to check that `*borrow` is live, but does not read the value.
`match *borrow {}` both checks that `*borrow` is live, and fake-reads the value.

Continuation of ~https://github.com/rust-lang/rust/pull/102256~ ~https://github.com/rust-lang/rust/pull/104844~

Fixes https://github.com/rust-lang/rust/issues/99180 https://github.com/rust-lang/rust/issues/53114

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/step.rs
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/step.rs:120]: IMO this should have semantics in Miri... but I guess that is for later?

Comment (cjgillot) [reply compiler/rustc_const_eval/src/interpret/step.rs:120]: I know way too little about miri to suggest semantics here.  There is room for later here.  Do you have an idea of what this should be?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/step.rs:120]: Something like
```rust
let _op = self.eval_place_to_op(place)?;
```

Comment (JakobDegen) [reply compiler/rustc_const_eval/src/interpret/step.rs:120]: The documentation in `mir/syntax.rs` would need to be changed to reflect this and it would need to be preserved throughout all mir dialects. So yeah, should probably be left for later

**Comment by scottmcm** [comment]
On the `let _: Ty = b;` example:

What happens there when it's more than just an ascription because it's a coercion site?

For example,
```
let b: Box<String> = …;
let _: Box<dyn Display> = b;
return b;
```

I guess that the coercion becomes an explicit statement in MIR, so that would still be rejected even though the `AscribeUserType` in MIR wouldn't be considered a use any more?

**Comment by eholk** [comment]
I'm probably not the best person to review this, so I'll re-roll.

r? @rust-lang/compiler

**Comment by oli-obk** [comment]
Triage: This is still waiting on https://github.com/rust-lang/rust/pull/102256

**Comment by RalfJung** [comment]
> The logical conclusion is that the match version should not check for unintialised places nor check that borrows are still live.

To be clear, if EXPR is `*bor` then `bor` must still be live for all three of these expressions, right?

**Comment by oli-obk** [comment]
The impl lgtm. I'm not sure we need second approval from T-lang, but I'll need to dig through the discussion before I'm sure about what everyone agreed to.

**Comment by cjgillot** [comment]
This is still blocked on https://github.com/rust-lang/rust/pull/104844

File: compiler/rustc_mir_build/src/build/matches/mod.rs
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:208]: The documentation of `lower_scrutinee` needs updating.

**Review by oli-obk — CHANGES_REQUESTED** [review]
I don't remember the outcome of these lang team meetings. Should we request a lang team FCP or was this already covered?

**Comment by RalfJung** [comment]
> match *borrow { _ => {} } continues to check that *borrow is live, but does not read the value.

Does this introduce a PlaceMention? That should be needed to fix Miri issue https://github.com/rust-lang/miri/issues/2360. If so, would be nice to see a test in this PR.

**Comment by oli-obk** [comment]
impl lgtm and the lang change makes sense. Is this something we'd let T-opsem do an FCP on?

**Comment by RalfJung** [comment]
There was the lang team meeting at https://github.com/rust-lang/lang-team/issues/184.  Has a summary of the results been posted somewhere?

Cc @rust-lang/opsem this is partially about borrow checking but also about opsem concerns, namely whether the two examples in https://github.com/rust-lang/miri/issues/2360 should be UB.

**Comment by JakobDegen** [comment]
Does this PR allow this code to compile:

```rust
let x: !;
match x {
    _ if {
        println!("Executed match!");
        loop {}
    } => (),
}
```

If I understand the description correctly then it does. This becomes slightly concerning in combination with #110105 , where the claim is specifically that such code must always be UB

**Comment by RalfJung** [comment]
I would not expect that code to compile since `x` has not been initialized. So in MIR I expect there to be no StorageLive for `x` making even just the place evaluation of `x` be UB.

But I'm not sure if that is how things actually work...

**Comment by cjgillot** [comment]
Actually, I corrected the test: if I use a `enum Void {}` instead of `!`, it compiles. Presumably because typeck confuses `!` with `()` somewhere.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, and we agreed that this should be delegated entirely to @rust-lang/opsem, who should decide this via FCP.

**Comment by RalfJung** [comment]
> Actually, I corrected the test: if I use a `enum Void {}` instead of `!`, it compiles. Presumably because typeck confuses `!` with `()` somewhere.

That's a bit strange, no? How does it ever consider a local of type `Void` to be live?

**Comment by RalfJung** [comment]
Sorry, somehow this fell of my radar, I didn't realize this was waiting for T-opsem. I'll create a meeting proposal.

**Comment by RalfJung** [comment]
@cjgillot [this question](https://github.com/rust-lang/rust/pull/103208#issuecomment-1606962172) is still open. Can you explain why it accepts that example?

**Comment by cjgillot** [comment]
> @cjgillot [this question](https://github.com/rust-lang/rust/pull/103208#issuecomment-1606962172) is still open. Can you explain why it accepts that example?

In short: `x` has storage starting from the `let` statement, until return. So it has storage at the scrutinee, even if uninitialized.

```
enum Void {}

fn main() {
    let x: Void;
    match x {
        _ if { loop {} } => (),
    }
}
```

The built MIR is pretty much:
```
debug x => _1;
StorageLive(_1); // `let x: Void;`
PlaceMention(_1); // as match scrutinee. Ok since `_1` has storage
goto bb1

bb1: // the `_` arm
goto bb1 // the `loop {}`

bb2: // the match guard, `_2` is the condition
SwitchInt(_2) -> [0: bb3; otherwise: bb4]

bb3: // the fallback arm
FakeRead(ForMatchedPlace(None), _1);
unreachable

bb4: // the return block
_0 = ()
StorageDead(_1)
return
```

The `PlaceMention(_1)` succeeds as `_1` has storage. The `FakeRead(_1)` would fail, but is never reached: the `_` arm is always taken, and we never get out of the guard.

Note: Using `!` is broken due  to never-type specialness. https://github.com/rust-lang/rust/issues/115716. Using `never_type_fallback` or `Void` gives the same MIR.

**Comment by RalfJung** [comment]
> x has storage starting from the let statement, until return

Ah, that makes sense, thanks!

**Comment by RalfJung** [comment]
Another question, regarding this desugaring:
```
PlaceMention(x); // for uniformity, to avoid losing information
match x {
  PATTERN => ARM,
  // compiler inserted fallback pattern
  _ => {
    FakeRead(ForMatchedPlace(None), x); // check that `x` is initialized here
    Unreachable;
  }
}
```
We ensure that matches are exhaustive, so I'm a bit surprised the fallback arm even makes any difference. Is this literally what happens or is this an approximation?

And the purpose of this, I supposed, is to reject code like this?

```rust
enum Void {}
let x: (i32, Void);
match x.1 {}
```

Here we do need to have a read from `x.1` at some point. In my head this is justified by implicitly inserted "never patterns" but those don't exist in Rust so I guess we'll have to make do with FakeRead in the mean time.

I was concerned about this read being fake and hence Miri missing the UB, but you have an `Unreachable` after it so Miri would still see the UB. Do we have a test ensuring this is detected as UB?
```rust
enum Void {}
let ptr: *const (i32, Void) = &0i32 as *const i32 as *const _;
unsafe { match (*ptr).1 {} }
```

This means there is a very strong requirement that this FakeRead can only ever be reached if the type is uninhabited.

**Comment by cjgillot** [comment]
> We ensure that matches are exhaustive, so I'm a bit surprised the fallback arm even makes any difference. Does the borrow checker realize that this is unreachable when there is a `_` pattern before? What about partially-`_` patterns? Is this literally what happens or is this an approximation?

For `_` patterns and partially-`_` patterns, or more generally for irrefutable patterns, the `FakeRead` will be unreachable in the MIR CFG, with literally no incoming edge.

If there is no irrefutable pattern, stuff becomes more interesting.

Match exhaustiveness pass is purely type-based and takes into account inhabitedness. (If `exhaustive_patterns` feature is active, or for the famous "match on !" pattern.)

In that case, match checking verifies that the pattern is "exhaustive accounting inhabitedness", but it may be "non-exhaustive ignoring inhabitedness". The difference is a set of uninhabited patterns that will jump into the inserted `FakeRead` arm, so borrowck can ensure that everything is initialized.

> I was concerned about this read being fake and hence Miri missing the UB, but you have an Unreachable after it so Miri would still see the UB. Do we have a test ensuring this is detected as UB?

Added a miri test.

**Comment by RalfJung** [comment]
Thanks for the explanation!

I think we might already have such a test in `tests/fail/never_transmute_void.rs`, but your test is sufficiently different that I like having both.

I assume we already have a test ensuring that this is rejected by the compiler?
```rust
enum Void {}
let x: Void;
match x {}
```

**Comment by cjgillot** [comment]
Yes. This one. https://github.com/rust-lang/rust/blob/7418413a7fad1c4e8b82f970bd78af030e5f813e/tests/ui/nll/match-on-borrowed.rs#L91

**Comment by RalfJung** [comment]
The test filename and comment don't really reflect what `never_init` tests for... but I'm happy to see the test, I think I understand now what is going on. Thanks!

**Comment by RalfJung** [comment]
In the t-opsem meeting, everyone was fine with the opsem implications of this change.
@rfcbot fcp merge

One point that came up is that the `FakeRead; Unreachable` in the desugaring (in the `_` fallback arm for uninhabited variants) could actually be `Read; Unreachable`. Having a real read is fine there and might even lead to better errors in Miri since the error will have something to do with the type being matched on, rather than just saying "entering unreachable code". A fake read is always somewhat suspicious; why would we want to "hide" the read from Miri / opsem?

**Comment by RalfJung** [comment]
What does this PR do for examples like this?
```rust
fn main() {
    #[derive(Copy, Clone)]
    enum Void {}
    union Uninit<T: Copy> {
        value: T,
        uninit: (),
    }
    unsafe {
        let x: Uninit<Void> = Uninit { uninit: () };
        match x.value {
            _ => println!("hi from the void!"),
        }
    }
}
```
I assume they are well-defined, i.e. Miri does not report UB? Do we have a test for that?

**Comment by RalfJung** [comment]
The Miri tests LGTM. @oli-obk wrote

> The impl lgtm.

but that was in May. Oli if you could briefly skim this again to make sure it still looks good to you that would be great. :)

**Comment by oli-obk** [comment]
Did another review of the impl and mir opt test changes. Nothing in mir or mir building has changed that would affect these changes, and they look good to me (I don't remember them from the last review, so this was a review from scratch 😅)

@bors r=oli-obk,RalfJung

**Comment by RalfJung** [comment]
It took more than a year, but it's finally done. :) Thanks a lot @cjgillot for seeing this through to the end!

**Comment by WaffleLapkin** [comment]
Should we add https://github.com/rust-lang/rust/labels/relnotes here? Might be useful to include in the release notes that we now accept more code

**Comment by RalfJung** [comment]
Sure, why not.

**Pull request merged**
